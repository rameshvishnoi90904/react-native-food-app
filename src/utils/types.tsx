export type FoodItem = {
    id: number,
    label: string,
    description: string,
    displayOrder: number,
    unitPriceFractional: number,
    currency: string,
    imageUrl: string,
    itemStock: {
        quantityLeft: number
    }
}

export type InnerSectionList = {
    sectionId: string,
    sectionDescription: string,
    disabledReason: string,
    list: FoodItem[],
    subSection: SectionItem[]
}

export type SectionItem = {
    id: number,
    label: string,
    description: string,
    displayOrder: number,
    disabled: Boolean,
    disabledReason: string|null,
    data: InnerSectionList[],
    outerSectionTitle: string
}


export type NavigationScreenPropsType = {
    Home: undefined;
    Detail: {item: FoodItem, disabledSection: boolean};
};

export interface CartState {
    [key: number]: CartItem;
}

export type CartItem = {
    foodId: number,
    quantity: number
}

export type dispatchActionType = {
    type: string;
    foodId: number;
}
