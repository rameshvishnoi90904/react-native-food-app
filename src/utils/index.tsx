import { FoodItem, SectionItem } from "./types";
/**
 * 
 * @param item section item
 * @param sectionTitle outer nested section title
 * @returns section item with outer nested section title
 */
const formatInnerSection = (item: any, sectionTitle: string) => {
    const foodItems = [...item.items];
    foodItems.sort((a: FoodItem, b: FoodItem) => (a.displayOrder - b.displayOrder));

    const toReturn: SectionItem = {
        id: item.id, 
        label: item.label,
        outerSectionTitle: sectionTitle,
        description: item.description,
        disabled: item.disabled,
        disabledReason: item.disabledReason,
        displayOrder: item.displayOrder,
        data: [{list: item.items, sectionId: item.id, sectionDescription: item.description, disabledReason: item.disabled ? item.disabledReason : "", subSection: []}],
    };
    return toReturn;
}
/**
 * 
 * @param item section data from api
 * @returns section item compatible with react native SectionList component
 */
export const formatSection = (item: any) : SectionItem => {
    const foodItems = [...item.items];
    const subSections = [...item.subSections];

    foodItems.sort((a: FoodItem, b: FoodItem) => (a.displayOrder - b.displayOrder));
    subSections.sort((a: SectionItem, b: SectionItem) => (a.displayOrder - b.displayOrder));
    
    const formattedSubSection: SectionItem[] = subSections.map((sItem) => formatInnerSection(sItem, item.label))

    const toReturn = {
        id: item.id, 
        label: item.label,
        description: item.description,
        disabled: item.disabled,
        outerSectionTitle: "",
        disabledReason: item.disabledReason,
        displayOrder: item.displayOrder,
        data: [{list: item.items, sectionId: item.id, sectionDescription: item.description, disabledReason: item.disabled ? item.disabledReason : "", subSection: formattedSubSection}],
    };
    return toReturn;
}