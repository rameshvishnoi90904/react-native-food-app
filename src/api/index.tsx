import { formatSection } from "../utils";
import { SectionItem } from "../utils/types";
export const fetchMenu = () : Promise<SectionItem[]> => {
    const url = 'https://atlas-fe-menu.atlas-kitchen.workers.dev/menu'
    return fetch(url)
        .then(response => response.json())
        .then(value => {
            const sections  = value.sections.map((item: any) : SectionItem => formatSection(item));

            sections.sort((a: SectionItem, b: SectionItem) => (a.displayOrder - b.displayOrder))

            return sections;
    })
    .catch(error => {
        throw error;
    });
}