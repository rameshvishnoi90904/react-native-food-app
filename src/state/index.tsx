
import { createContext } from 'react';
import { dispatchActionType, CartState, SectionItem } from '../utils/types';

export type AppContextType = {
    state: {cart: CartState, foodSection: SectionItem[]};
    dispatch: (action: dispatchActionType) => void;
}

export const AppContext = createContext<AppContextType | null>(null);