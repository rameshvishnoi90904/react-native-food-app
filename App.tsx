/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 */

import React, { useEffect, useState } from 'react';
import {
  SafeAreaView,
} from 'react-native';
import MenuScreen from './src/screens/MenuScreen';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import {CartState, NavigationScreenPropsType, SectionItem, dispatchActionType} from './src/utils/types';
import DetailScreen from './src/screens/DetailScreen';
import { AppContext } from './src/state';
import { fetchMenu } from './src/api';
const Stack = createNativeStackNavigator<NavigationScreenPropsType>();

function App(): React.JSX.Element {
  const [cart, updateCart] = useState<CartState>({})
  const [foodSection, setFoodSection] = useState<SectionItem[]>([])

  useEffect(() => {
    fetchMenu().then((value: SectionItem[]) => {
      setFoodSection(value)
  })
  },[])

  const dispatch = (action: dispatchActionType) => {
    const toReturn = {...cart}

    switch (action.type) {
      case 'add': 
        if (toReturn[action.foodId]) {
          toReturn[action.foodId].quantity++
        } else {
          toReturn[action.foodId] = {
            quantity: 1,
            foodId: action.foodId
          }
        }
      break;
      case 'remove': 
        toReturn[action.foodId].quantity--
      break;
    }
    updateCart(toReturn)
  }
  const state = {cart: cart, foodSection: foodSection}
  return (
    <SafeAreaView style={{flex: 1}}>
      <AppContext.Provider value={{state, dispatch}}>
        <NavigationContainer>
          <Stack.Navigator>
            <Stack.Screen name="Home" component={MenuScreen} />
            <Stack.Screen name="Detail" component={DetailScreen} />
          </Stack.Navigator>
        </NavigationContainer>
      </AppContext.Provider>
    </SafeAreaView>
  );
}

export default App;
