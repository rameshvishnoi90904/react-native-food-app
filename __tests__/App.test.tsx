import 'react-native';
import React from 'react';
import App from '../App';
import { render, act, waitFor, fireEvent } from '@testing-library/react-native';

// Note: import explicitly to use the types shipped with jest.
import {it, expect, jest} from '@jest/globals';

import mockData from '../src/api/fakeData.json'
import { SectionItem } from '../src/utils/types';
import { formatSection } from '../src/utils';

const mockSections  = mockData.sections.map((item: any) : SectionItem => formatSection(item));

mockSections.sort((a: SectionItem, b: SectionItem) => (a.displayOrder - b.displayOrder))

jest.mock('../src/api/', () => ({
  fetchMenu: () => new Promise((resolve, reject) => {
    resolve([...mockSections])
  })
}))


it('renders correctly', async () => {
  const tree = render(<App />).toJSON();
  await waitFor(() => {
    expect(tree).toBeDefined();
  })
  expect(tree).toMatchSnapshot(); 
});

it('test dispatch actions', async () => {
  const firstFoodItem = mockSections[0].data[0].list[0];

  const screen = render(<App />);
  await waitFor(() => {
    expect(screen).toBeDefined();
  })
  

  const firstFoodItemEle = await screen.findByTestId(`add-cart-button-${firstFoodItem.id}`);

  act(() => {
    fireEvent.press(firstFoodItemEle)
  })

  const incrementCart = await screen.findByTestId(`add-cart-${firstFoodItem.id}`);

  act(() => {
    fireEvent.press(incrementCart)
  })


  expect(screen.getByTestId(`quantity-cart-${firstFoodItem.id}`).children).toEqual(["2"])


  const decrementCart = await screen.findByTestId(`remove-cart-${firstFoodItem.id}`);

  act(() => {
    fireEvent.press(decrementCart)
  })

  expect(screen.getByTestId(`quantity-cart-${firstFoodItem.id}`).children).toEqual(["1"])

});

