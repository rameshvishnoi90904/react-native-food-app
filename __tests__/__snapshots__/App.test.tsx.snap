// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenStack
      style={
        {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        collapsable={false}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        nativeBackButtonDismissalEnabled={false}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onGestureCancel={[Function]}
        onHeaderBackButtonClicked={[Function]}
        onNativeDismissCancelled={[Function]}
        onTransitionProgress={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        sheetAllowedDetents="large"
        sheetCornerRadius={-1}
        sheetExpandsWhenScrolledToEdge={true}
        sheetGrabberVisible={false}
        sheetLargestUndimmedDetent="all"
        stackPresentation="push"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        swipeDirection="horizontal"
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            {
              "flex": 1,
              "flexDirection": "column-reverse",
            }
          }
        >
          <View
            collapsable={false}
            style={
              [
                {
                  "flex": 1,
                },
                {
                  "backgroundColor": "rgb(242, 242, 242)",
                },
                undefined,
              ]
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "rgb(255, 254, 250)",
                  "flex": 1,
                }
              }
            >
              <RCTScrollView
                data={[]}
                getItem={[Function]}
                getItemCount={[Function]}
                keyExtractor={[Function]}
                onContentSizeChange={[Function]}
                onLayout={[Function]}
                onMomentumScrollBegin={[Function]}
                onMomentumScrollEnd={[Function]}
                onScroll={[Function]}
                onScrollBeginDrag={[Function]}
                onScrollEndDrag={[Function]}
                renderItem={[Function]}
                scrollEventThrottle={0.0001}
                stickyHeaderIndices={[]}
              >
                <View />
              </RCTScrollView>
            </View>
          </View>
        </View>
        <RNSScreenStackHeaderConfig
          backButtonInCustomView={false}
          backTitleVisible={true}
          backgroundColor="rgb(255, 255, 255)"
          color="rgb(0, 122, 255)"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={false}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="Home"
          titleColor="rgb(28, 28, 30)"
          topInsetEnabled={false}
          translucent={false}
        />
      </RNSScreen>
    </RNSScreenStack>
  </RNCSafeAreaProvider>
</RCTSafeAreaView>
`;
